stellargraph/data/explorer.py

```diff
@@ -1021,93 +755,137 @@ def _check_parameter_values(self, nodes, n, n_size, graph_schema, seed):
+
+                [[node]
+                 [in_1...in_n]  [out_1...out_m]
+                 [in_1.in_1...in_n.in_p] [in_1.out_1...in_n.out_q]
+                    [out_1.in_1...out_m.in_p] [out_1.out_1...out_m.out_q]
+                 [in_1.in_1.in_1...in_n.in_p.in_r] [in_1.in_1.out_1...in_n.in_p.out_s] ...
+                 ...]
+
+            where a single, undirected walk might be, for example:
+
+                [node out_i  out_i.in_j  out_i.in_j.in_k ...]
+        """
+        rs = self._check_parameter_values(nodes, n, in_size, out_size, seed)
+        max_hops = len(in_size)
+        max_slots = 2 ** (max_hops + 1) - 1
+
+        samples = []
+
+        for node in nodes:  # iterate over root nodes
+            for _ in range(n):  # do n bounded breadth first walks from each root node
+                q = list()  # the queue of neighbours
+                # the list of sampled node-lists:
+                sample = [[] for _ in range(max_slots)]
+                # Add node to queue as (node, depth, slot)
+                q.append((node, 0, 0))
+
+                while len(q) > 0:
+                    # remove the top element in the queue
+                    # index 0 pop the item from the front of the list
+                    cur_node, cur_depth, cur_slot = q.pop(0)
+                    sample[cur_slot].append(cur_node)  # add to the walk
+                    depth = cur_depth + 1  # the depth of the neighbouring nodes
+
+                    # consider the subgraph up to and including max_hops from root node
+                    if depth > max_hops:
+                        continue
+                    # get in-nodes
+                    neighbours = self._sample_neighbours(
+                        rs, cur_node, 0, in_size[cur_depth]
                     )
-                )
-            if seed < 0:
-                raise ValueError(
-                    "({}) The random number generator seed value, seed, should be positive integer or None.".format(
-                        type(self).__name__
+                    # add them to the back of the queue
+                    slot = 2 * cur_slot + 1
+                    q.extend(
+                        [(sampled_node, depth, slot) for sampled_node in neighbours]
                     )
-                )
+                    # get out-nodes
+                    neighbours = self._sample_neighbours(
+                        rs, cur_node, 1, out_size[cur_depth]
+                    )
+                    # add them to the back of the queue
+                    slot = slot + 1
+                    q.extend(
+                        [(sampled_node, depth, slot) for sampled_node in neighbours]
+                    )
+
+                # finished multi-hop neighbourhood sampling
+                samples.append(sample)
+
+        return samples
+
+    def _sample_neighbours(self, rs, node, idx, size):
```

Comment:
This is a nice general method. Do you think it should replace `GraphWalk.neighbors`?
Also, could you please add a docstring explaining the args, especially `idx`? It's clear now, but I'm sure we can forget in the future.

Reply:
I think `in_edges` and `out_edges` methods only exist in DiGraph.

Reply:
We can either check if the graph is directed inside this method (the easier solution?), or make all graphs directed.

Reply:
My preference is that all graphs are directed by default, and the `neighbors()` method gives back nodes from both in and out neighbourhoods. Alternatively, for an undirected graph we could add `in_edges()` and `out_edges()` methods.